bonsai/models/umt5/tests/test_umt5.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2025 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test accuracy of jax impelement vs torch transformer impelement."""
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+import torch
+from absl.testing import absltest
+from huggingface_hub import snapshot_download
+from transformers import AutoTokenizer
+from transformers import UMT5EncoderModel as TorchUMT5EncoderModel
+from transformers import UMT5Model as TorchUMT5Model
+
+from bonsai.models.umt5.modeling import UMT5Config, UMT5EncoderModel, UMT5Model
+from bonsai.models.umt5.params import create_model, load_model_config
+
+
+def compare_outputs(jax_output: jax.Array, torch_output, name: str, rtol: float = 1e-3, atol: float = 1e-5):
+    """Compare JAX and PyTorch outputs and report differences.
+    Args:
+        jax_output: Output from JAX model
+        torch_output: Output from PyTorch model (torch.Tensor)
+        name: Name of the output being compared
+        rtol: Relative tolerance
+        atol: Absolute tolerance
+    """
+    if torch_output.dtype == torch.bfloat16:
+        torch_output = torch_output.float()
+    if jax_output.dtype == jnp.bfloat16:
+        jax_output = jax_output.astype(jnp.float32)
+
+    # Convert PyTorch to numpy
+    if isinstance(torch_output, torch.Tensor):
+        torch_np = torch_output.detach().cpu().numpy()
+    else:
+        torch_np = np.array(torch_output)
+
+    # Convert JAX to numpy
+    jax_np = np.array(jax_output)
+
+    # Check shapes match
+    if jax_np.shape != torch_np.shape:
+        print(f"❌ Shape mismatch! jax shape: {jax_np.shape}, torch shape: {torch_np.shape}")
+        return False
+
+    if jax_np.dtype != torch_np.dtype:
+        print(f"❌ Shape mismatch! jax dtype: {jax_np.dtype}, torch dtype: {torch_np.dtype}")
+        return False
+
+    # Check if within tolerance
+    close = np.allclose(jax_np, torch_np, rtol=rtol, atol=atol)
+
+    if not close:
+        print(f"\n❌ Outputs do NOT match (rtol={rtol}, atol={atol})")
+        # Show some mismatched locations
+        mismatch_mask = ~np.isclose(jax_np, torch_np, rtol=rtol, atol=atol)
+        n_mismatches = np.sum(mismatch_mask)
+        print(f"  Number of mismatches: {n_mismatches} / {jax_np.size} ({100 * n_mismatches / jax_np.size:.2f}%)")
+
+    return close
```

Comment:
Let's make this dense as possible. i.e:


```suggestion
def to_np(array) -> np.ndarray:
    """Standardizes JAX/Torch/NumPy arrays to float32 NumPy."""
    if hasattr(array, "detach"):  # Torch path
        return array.detach().cpu().float().numpy()
    return np.array(array, dtype=np.float32)  # JAX/NumPy path

def compare_outputs(jax_output, torch_output, name: str, rtol=1e-3, atol=1e-5):
    jax_np = to_np(jax_output)
    torch_np = to_np(torch_output)

    try:
        np.testing.assert_allclose(jax_np, torch_np, rtol=rtol, atol=atol)
        return True

    except AssertionError as e:
        print(f"\nFAIL: {name} comparison failed")
        print(e)
        return False
```